File: csharp/src/Projection.cs
namespace Volante
{
    using System;
    using System.Collections;
    using System.Collections.Generic;
    using System.Reflection;

    /// <summary>
    /// Class used to project selected objects using relation field. 
    /// For all selected objects (specified by array or iterator), 
    /// value of specified field (of IPersistent, array of IPersistent, Link or Relation type)
    /// is inspected and all referenced object for projection (duplicate values are eliminated)
    /// </summary>
    public class Projection<From, To> : ICollection<To>
        where From : class,IPersistent
        where To : class,IPersistent
    {
        /// <summary>
        /// Constructor of projection specified by field name of projected objects
        /// </summary>
        /// <param name="fieldName">field name used to perform projection</param>
        public Projection(String fieldName)
        {
            SetProjectionField(fieldName);
        }

        /// <summary>
        /// Default constructor of projection. This constructor should be used
        /// only when you are going to derive your class from Projection and redefine
        /// Map() method in it or sepcify type and fieldName later using SetProjectionField()
        /// method
        /// </summary>
        public Projection() { }

        public int Count
        {
            get
            {
                return hash.Count;
            }
        }

        public bool IsSynchronized
        {
            get
            {
                return false;
            }
        }

        public object SyncRoot
        {
            get
            {
                return null;
            }
        }

        public void CopyTo(To[] dst, int i)
        {
            foreach (object o in this)
            {
                dst.SetValue(o, i++);
            }
        }

        /// <summary>
        /// Specify projection field name
        /// </summary>
        /// <param name="fieldName">field name used to perform projection</param>
        public void SetProjectionField(string fieldName)
        {
            Type type = typeof(From);
            field = type.GetField(fieldName, BindingFlags.Instance | BindingFlags.NonPublic | BindingFlags.Public);
            if (field == null)
                throw new DatabaseException(DatabaseException.ErrorCode.KEY_NOT_FOUND);
        }

        /// <summary>
        /// Project specified selection
        /// </summary>
        /// <param name="selection">array with selected object</param>
        public void Project(From[] selection)
        {
            for (int i = 0; i < selection.Length; i++)
            {
                Map(selection[i]);
            }
        }

        /// <summary>
        /// Project specified object
        /// </summary>
        /// <param name="obj">selected object</param>
        public void Project(From obj)
        {
            Map(obj);
        }

        /// <summary>
        /// Project specified selection
        /// </summary>
        /// <param name="selection">enumerator specifying selceted objects</param>
        public void Project(IEnumerator<From> selection)
        {
            while (selection.MoveNext())
            {
                Map(selection.Current);
            }
        }

        /// <summary>
        /// Project specified selection
        /// </summary>
        /// <param name="selection">enumerator specifying selceted objects</param>
        public void Project(IEnumerable<From> selection)
        {
            foreach (From obj in selection)
            {
                Map(obj);
            }
        }

        /// <summary>
        /// Join this projection with another projection.
        /// Result of this join is set of objects present in both projections.
        /// </summary>
        /// <param name="prj">joined projection</param>
        public void Join<X>(Projection<X, To> prj) where X : class,IPersistent
        {
            Dictionary<To, To> join = new Dictionary<To, To>();
            foreach (To p in prj.hash.Keys)
            {
                if (hash.ContainsKey(p))
                    join[p] = p;
            }
            hash = join;
        }

        /// <summary>
        /// Get result of preceding project and join operations
        /// </summary>
        /// <returns>array of objects</returns>
        public To[] ToArray()
        {
            To[] arr = new To[hash.Count];
            hash.Keys.CopyTo(arr, 0);
            return arr;
        }

        /// <summary>
        /// Get number of objets in the result 
        /// </summary>
        public int Length
        {
            get
            {
                return hash.Count;
            }
        }

        /// <summary>
        /// Get enumerator for the result of preceding project and join operations
        /// </summary>
        /// <returns>enumerator</returns>
        public IEnumerator<To> GetEnumerator()
        {
            return hash.Keys.GetEnumerator();
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return GetEnumerator();
        }

        /// <summary>
        /// Reset projection - clear result of prceding project and join operations
        /// </summary>
        public void Reset()
        {
            hash.Clear();
        }

        /// <summary>
        /// Add object to the set
        /// </summary>
        /// <param name="obj">object to be added to the set</param>
        public void Add(To obj)
        {
            if (obj != null)
                hash[obj] = obj;
        }

        /// <summary>
        /// Get related objects for the object obj. 
        /// It's possible to redefine this method in derived classes 
        /// to provide application specific mapping
        /// </summary>
        /// <param name="obj">object from the selection</param>
        protected void Map(From obj)
        {
            if (field == null)
            {
                Add((To)(object)obj);
                return;
            }

            object o = field.GetValue(obj);
            if (o is ILink<To>)
            {
                To[] arr = ((ILink<To>)o).ToArray();
                for (int i = 0; i < arr.Length; i++)
                {
                    Add(arr[i]);
                }
                return;
            }

            if (o is To[])
            {
                To[] arr = (To[])o;
                for (int i = 0; i < arr.Length; i++)
                {
                    Add(arr[i]);
                }
                return;
            }

            Add((To)o);
        }

        public bool IsReadOnly
        {
            get
            {
                return false;
            }
        }

        public bool Contains(To obj)
        {
            return hash.ContainsKey(obj);
        }

        public bool Remove(To obj)
        {
            return hash.Remove(obj);
        }

        public void Clear()
        {
            hash.Clear();
        }

        private Dictionary<To, To> hash = new Dictionary<To, To>();
        private FieldInfo field;
    }
}

File: java/src15/org/nachodb/Projection.java
package org.nachodb;

import java.util.*;
import java.lang.reflect.Field;

/**
 * Class use to project selected objects using relation field. 
 * For all selected objects (specified by array ort iterator), 
 * value of specified field (of IPersistent, array of IPersistent, Link or Relation type)
 * is inspected and all referenced object for projection (duplicate values are eliminated)
 */
public class Projection<From extends IPersistent, To extends IPersistent> extends HashSet<To> { 
    /**
     * Constructor of projection specified by class and field name of projected objects
     * @param type base class for selected objects
     * @param fieldName field name used to perform projection
     */
    public Projection(Class type, String fieldName) { 
        setProjectionField(type, fieldName);
    }

    /**
     * Default constructor of projection. This constructor should be used
     * only when you are going to derive your class from Projection and redefine
     * map method in it or specify type and fieldName later using setProjectionField
     * method
     */
    public Projection() {}

    /** 
     * Specify class of the projected objects and projection field name
     * @param type base class for selected objects
     * @param fieldName field name used to perform projection
     */
    public void setProjectionField(Class type, String fieldName) { 
        try { 
            field = type.getDeclaredField(fieldName);
            field.setAccessible(true);
        } catch (Exception x) { 
            throw new StorageError(StorageError.KEY_NOT_FOUND, x);
        }
    }

    /**
     * Project specified selection
     * @param selection array with selected object
     */
    public void project(From[] selection) { 
        for (int i = 0; i < selection.length; i++) { 
            map(selection[i]);
        }
    } 

    /**
     * Project specified object
     * @param obj selected object
     */
    public void project(From obj) { 
        map(obj);
    } 

    /**
     * Project specified selection
     * @param selection iterator specifying selected objects
     */
    public void project(Iterator<From> selection) { 
        while (selection.hasNext()) { 
            map(selection.next());
        }
    } 

    /**
     * Project specified selection
     * @param c selection iterator specifying selected objects
     */
    public void project(Collection<From> c) { 
        for (From o : c) { 
            map(o);
        }
    } 

    /**
     * Join this projection with another projection.
     * Result of this join is set of objects present in both projections.
     */
    public void join(Projection<From, To> prj) { 
        retainAll(prj);
    }

    /**
     * Get result of preceding project and join operations
     * @return array of objects
     */
    public IPersistent[] toPersistentArray() { 
        return (IPersistent[])toArray(new IPersistent[size()]);
    }

    /**
     * Reset projection - clear result of preceding project and join operations
     */
    public void reset() { 
        clear();
    }

    /**
     * Add object to the set
     * @param obj objet to be added
     */
    public boolean add(To obj) { 
        if (obj != null) { 
            return super.add(obj);
        }
        return false;
    }

    /**
     * Get related objects for the object obj. 
     * It is possible to redifine this method in derived classes 
     * to provide application specific mapping
     * @param obj object from the selection
     */
    protected void map(From obj) {   
        if (field == null) { 
            add((To)obj);
        } else { 
            try { 
                Object o = field.get(obj);
                if (o instanceof Link) { 
                    Object[] arr = ((Link)o).toArray();
                    for (int i = 0; i < arr.length; i++) { 
                        add((To)arr[i]);
                    }
                } else if (o instanceof Object[]) { 
                    Object[] arr = (Object[])o;
                    for (int i = 0; i < arr.length; i++) { 
                        add((To)arr[i]);
                    }
                } else { 
                    add((To)o);
                }
            } catch (Exception x) { 
                throw new StorageError(StorageError.ACCESS_VIOLATION, x);
            }
        }
    }

    private Field field;
}
